Install: Add use-buildcache option to install

kwryankrattiger opened the pull request:
Allow differentiating between top level packages and dependencies when determining whether to install from the cache or not.

@scottwittenburg 

Implements this [feature request](https://github.com/spack/spack/issues/29695).

kwryankrattiger committed:
Install: Add use-buildcache option to install

Allow differentiating between top level packages and dependencies when
determining whether to install from the cache or not.

kwryankrattiger commented:
@spackbot fix style

kwryankrattiger committed:
Add unit test for --use-buildcache

scottwittenburg commented:
@kwryankrattiger Can you rebase this on the latest `develop` and force-push?  I'm hoping it will then get past the failing checks we're seeing above.

scottwittenburg commented:
@spackbot fix style

kwryankrattiger commented on lib/spack/spack/cmd/install.py:
```diff
@@ -31,10 +32,64 @@
+    if "" in bc_map:
+        package_use_buildcache = bc_map[""]
+        dependencies_use_buildcache = bc_map[""]
+    if "package" in bc_map:
```
This will permit passing `--use-buildcache auto,package:never` and use the auto for dependencies. It is of a shorthand because I for one never type dependencies correctly and it is a long keyword.

scottwittenburg replied:
Is the idea here that you could also say `--use-buildcache auto,dependencies:never`?  I think it's weird if it only works the other way.

kwryankrattiger replied:
Both ways work, also not listing a "default" and only specifying one works.

so...
`--use-buildcache package=never` is the same as `--use-buildcache auto,package=never` or `--use-buildcache dependencies=auto,package=never`

or you could do
`--use-buildcache dependencies=only` which is the same as `--use-buildcache auto,dependencies=only` or `--use-buildcache package=auto,dependencies=only`

and so on. The short is, by default the option is "auto", or it is the specified default (ie. specified in `--use-buildcache` without a prefix), or it is overridden by a prefixed option, ie. `package:only` or `dependencies:never` etc.

I left it open to allow some flexibility on how to specify the option based on the way the user is reasoning on what they want. Not sure if that is correct or not, trying to toe the line between overly restrictive and annoying to use.

scottwittenburg commented on lib/spack/spack/cmd/install.py:
```diff
@@ -31,10 +32,64 @@
+            ":{[auto|only|never],[[package:[auto|only|never]],[dependencies:[auto|only|never]]]"
+        )
+        exit(1)
+        return None
```
Given the `exit(1)` just above this, I wouldn't think this `return` statement could ever be executed.

kwryankrattiger replied:
ah, that isn't supposed to exit(1). I wasn't sure how to handle errors correctly with parsing.

scottwittenburg replied:
Normally, from spack commands, you can just return the value you want to be used as the exit code.  Probably worth verifying that value it makes it all the way back to the shell though.  Unless your test method already validates the exit code, in which case it will let you know.

scottwittenburg commented on lib/spack/spack/cmd/install.py:
```diff
@@ -31,10 +32,64 @@
+    bc_parsed = {}
+    bc_keys = ["package", "dependencies", ""]
+    bc_values = ["only", "never", "auto"]
+    kv_list = re.findall("([a-z]+)?:?([a-z]+)", args.use_buildcache)
```
I'm not sure this regex is doing what we want.  If I say:

```
spack install --no-check-signature --use-buildcache auto readline
```

then spack errors with:

```
==> Error: Unrecognized arguments passed to use-buildcache: [('aut', 'o')]
==> Error: Expected: --use-buildcache :{[auto|only|never],[[package:[auto|only|never]],[dependencies:[auto|only|never]]]
```

kwryankrattiger replied:
I love regex! That is supposed to evaluate to `("", "auto")`, which it does for `never` and `only`...

kwryankrattiger replied:
`([a-z]+:)?([a-z]+)` works, but requires a more complicated filtering

scottwittenburg reviewed (changes requested):
Thanks @kwryankrattiger,!  This is looking pretty good, but I left a comment or two to to be addressed, along with a question about how the shorthand is intended to work. 

Also, could you provide a brief description of the change?  Even just a pointer to the issue this addresses could help other reviewers know what they're looking at

scottwittenburg commented:
I don't mind running the style fix for you, but you should at least pull the changes and incorporate them before you force push again :stuck_out_tongue_winking_eye:

scottwittenburg commented:
@spackbot fix style

kwryankrattiger committed:
Add test of supported edge case and bug fix

scottwittenburg commented on lib/spack/spack/cmd/install.py:
```diff
@@ -31,10 +32,61 @@
 level = "short"
 
 
+def parse_use_buildcache(args):
```
Can you add a docstring saying what `args` is?  In a lot of other places in spack commands, that variable name is used for the parsed args object, so I was sort of surprised it was just the value 0f the `--use-buildcache` option.  Also, the docstring should probably briefly describe the data structure returned to the caller.

scottwittenburg commented on lib/spack/spack/cmd/install.py:
```diff
@@ -123,6 +177,12 @@ def setup_parser(subparser):
+        "--use-buildcache",
+        dest="use_buildcache",
+        default="package:auto,dependencies:auto",
+        help="finer grain control over which packages to check mirrors for",
```
Is this the right place to describe a little more to the user about the expected format?  The `default` value does provide an example, but the values `only`, `never`, and `auto` are hidden from the argument parser in the `parse_use_buildcache()` method.  It's also probably worth mentioning (here, if this is the right place) the shorthand versions allowed.

scottwittenburg replied:
Maybe also worth mentioning that `auto` means "fall back to the (maybe default) value of` either `--use-cache` or `--cache-only`.

scottwittenburg commented on lib/spack/spack/cmd/install.py:
```diff
@@ -31,10 +32,61 @@
+    bc_map = {k: v for k, v in kv_list if k in bc_keys and v in bc_values}
+    if not len(kv_list) == len(bc_map):
+        unrecognized_args = [(k, v) for k, v in kv_list if k not in bc_keys or v not in bc_values]
+        tty.error("Unrecognized arguments passed to use-buildcache: %s" % unrecognized_args)
```
I kind of don't like the way the list of string tuples gets printed, e.g.:

```
$ spack install --use-buildcache always readline
==> Error: Unrecognized arguments passed to use-buildcache: [('', 'always')]
```

What do you think of omitting that bit from this particular line?   Since you print out the expected values below, it seems possible to discern what the problem is.

scottwittenburg reviewed (changes requested):
This is getting closer, thanks @kwryankrattiger!  It's working for my particular use case (must install deps from buildcache, never install package from buildcache), but I have a couple  requests regarding the UI.

Also, it looks like some unit tests failed, which this may or may not have caused.  Can you check if those failures are reproducible on your end?

kwryankrattiger committed:
Review fixes

kwryankrattiger committed:
Use metavar to display use-buildcache options.

scottwittenburg commented:
@spackbot fix style

kwryankrattiger committed:
Use install_mockery_mutable_config for fixture

kwryankrattiger committed:
Update spack-completion

kwryankrattiger committed:
Fix bug in distinguishing requested and explicit.

A requested package is not required to be "explicit"ly requested
when specifying --only package or --use-buildcache package:<...>
options in install. Add a task property that returns the former.

kwryankrattiger committed:
Fix installer unit tests

alalazo commented:
I have a high-level question. Failed pipeline jobs let you run the build in a docker image to inspect the failure. Within that Docker image there's an `install.sh` script which runs a single command to install the spec that failed. An example `install.sh` script is, for instance:
```bash
#!/bin/bash


# spack install command
"spack" "-d" "-v" "install" "--show-log-on-error" "--keep-stage" "--no-check-signature" "--cdash-upload-url" "https://cdash.spack.io/submit.php?project=Spack+Testing" "--cdash-build" "unifyfs@1.0%oneapi@2022.1.0 hash=36ftnpast6yvr3zawwnrcgxn5ao3t5zs arch=linux-ubuntu20.04-x86_64 (E4S OneAPI)" "--cdash-site" "Cloud Gitlab Infrastructure (uo-pegasus)" "--cdash-buildstamp" "20220919-0545-E4S OneAPI" "--no-add" "-f" "/builds/spack/spack/jobs_scratch_dir/reproduction/unifyfs.json"
```

I assume (let me know if this is a wrong assumption) that something similar to running `install.sh` happens in Gitlab. If the assumption is correct, why we discarded splitting a single `spack install` command into:
```console
$ spack install --cache-only --only=dependencies <spec>
$ spack install --no-cache --only=root <spec>
```
to ensure that dependencies are never installed from sources? Is there any technical issue that I overlooked proceeding that way?

scottwittenburg commented:
@alalazo We used to split it in two, but it complicated the ci module even more, as we had to make a copy of the `spack.yaml` before the first install, then use it to replace the version that spack clobbered before the next spack install.  

Also, the `install.sh` script you mentioned in your question, which ends up in the reproducer docker container, is the same one that gitlab runs in the pipeline.  Or did I misunderstand the question?

scottwittenburg commented:
By the way, in my pipelines testing "rebuild everything", this PR is working as expected.  @tgamblin was going to provide some insight on a better way, within the installer.py module, to distinguish dependencies from package.  But from my perspective, @kwryankrattiger figured that out in a way that seems reasonable to me.  So I'd be happy to accept this as-is, once the tests are passing.

alalazo commented:
> We used to split it in two, but it complicated the ci module even more, as we had to make a copy of the spack.yaml before the first install, then use it to replace the version that spack clobbered before the next spack install.

I don't think I get what you're describing above, which might be due to me being unfamiliar with `ci.py` and `cmd/ci.py`. Isn't this:

https://github.com/spack/spack/blob/0ddbb92ae3acad8b2b87198d127aa96ea95f48c0/lib/spack/spack/ci.py#L1985-L2005

the only part that eventually need to change? Based on the example above wouldn't it suffice to generate something similar to:
```bash
#!/bin/bash


# spack install command
"spack" "-d" "-v" "install" "--cache-only" "--only=dependencies" "--no-add" "-f" "/builds/spack/spack/jobs_scratch_dir/reproduction/unifyfs.json"
"spack" "-d" "-v" "install" "--show-log-on-error" "--no-cache" "--only=package" "--keep-stage" "--no-check-signature" "--cdash-upload-url" "https://cdash.spack.io/submit.php?project=Spack+Testing" "--cdash-build" "unifyfs@1.0%oneapi@2022.1.0 hash=36ftnpast6yvr3zawwnrcgxn5ao3t5zs arch=linux-ubuntu20.04-x86_64 (E4S OneAPI)" "--cdash-site" "Cloud Gitlab Infrastructure (uo-pegasus)" "--cdash-buildstamp" "20220919-0545-E4S OneAPI" "--no-add" "-f" "/builds/spack/spack/jobs_scratch_dir/reproduction/unifyfs.json"
```
?

scottwittenburg commented:
That might work, without having to make a backup copy of the environment and restore it between the installs, now that we have the `--no-add` option.  However, I would need to go back and make sure there wasn't some additional problem with the "two spack installs" approach.

kwryankrattiger committed:
Change name to is_root instead of is_package

scottwittenburg reviewed (approved):
I tested this with the other proposed changes we need for asking spackbot to rebuild everything, and it's working.

haampie reviewed (commented):
Seems like you're missing

```diff
diff --git a/lib/spack/spack/installer.py b/lib/spack/spack/installer.py
index 7ec31f0267..1ec21be45d 100644
--- a/lib/spack/spack/installer.py
+++ b/lib/spack/spack/installer.py
@@ -2391,7 +2391,7 @@ def get_deptypes(self, pkg):
         """
         deptypes = ["link", "run"]
         include_build_deps = self.install_args.get("include_build_deps")
-        if not self.install_args.get("cache_only") or include_build_deps:
+        if not self.install_args.get("package_cache_only") or include_build_deps:
             deptypes.append("build")
         if self.run_tests(pkg):
             deptypes.append("test")
```
an old reference to `cache_only`, which now causes the installer not to prune redundant build deps.
